scipy/optimize/_linprog_highs.py

```diff
@@ -0,0 +1,411 @@
+
+            nit : int
+                The total number of iterations performed.
+                For ``solver='simplex'``, this includes iterations in all
+                phases. For ``solver='ipm'``, this does not include
+                crossover iterations.
+            crossover_nit : int
+                The number of primal/dual pushes performed during the
+                crossover routine for ``solver='ipm'``.  This is ``0``
+                for ``solver='simplex'``.
+            message : str
+                A string descriptor of the exit status of the algorithm.
+
+    """
+
+    _check_unknown_options(unknown_options)
+
+    # Catch message_level warnings that HiGHS currently misses:
+    message_level = (
+        _check_invalid_option_values(message_level, 'message_level',
+                                     {None, 0, 1, 2, 3, 4, 5, 6, 7}, MESSAGE_LEVEL_MINIMAL))
+
+    # SIMPLEX_STRATEGY_PRIMAL (4) is experimental -- disallow use!
+    simplex_strategy = (
+        _check_invalid_option_values(simplex_strategy, 'simplex_strategy',
+                                     {None, 0, 1, 2, 3}, 1))
+
+    statuses = {
+        MODEL_STATUS_NOTSET: (4, 'HiGHS Status Code 0: HighsModelStatusNOTSET'),
+        MODEL_STATUS_LOAD_ERROR: (4, 'HiGHS Status Code 1: HighsModelStatusLOAD_ERROR'),
+        MODEL_STATUS_MODEL_ERROR: (2, 'HiGHS Status Code 2: HighsModelStatusMODEL_ERROR'),
+        MODEL_STATUS_MODEL_EMPTY: (4, 'HiGHS Status Code 3: HighsModelStatusMODEL_EMPTY'),
+        MODEL_STATUS_PRESOLVE_ERROR: (4, 'HiGHS Status Code 4: HighsModelStatusPRESOLVE_ERROR'),
+        MODEL_STATUS_SOLVE_ERROR: (4, 'HiGHS Status Code 5: HighsModelStatusSOLVE_ERROR'),
+        MODEL_STATUS_POSTSOLVE_ERROR: (4, 'HiGHS Status Code 6: HighsModelStatusPOSTSOLVE_ERROR'),
+        MODEL_STATUS_RDOVUB: (4, 'HiGHS Status Code 10: HighsModelStatusREACHED_DUAL_OBJECTIVE_VALUE_UPPER_BOUND'),
+        MODEL_STATUS_PRIMAL_INFEASIBLE: (2, "The problem is infeasible."),
+        MODEL_STATUS_PRIMAL_UNBOUNDED: (3, "The problem is unbounded."),
+        MODEL_STATUS_OPTIMAL: (0, "Optimization terminated successfully."),
+        MODEL_STATUS_REACHED_TIME_LIMIT: (1, "Time limit reached."),
+        MODEL_STATUS_REACHED_ITERATION_LIMIT: (1, "Iteration limit reached."),
+    }
+
+    c, A_ub, b_ub, A_eq, b_eq, bounds, x0 = lp
+
+    lb, ub = bounds.T.copy()                # separate bounds, copy->C-cntgs
+    # highs_wrapper solves LHS <= A*x <= RHS, not equality constraints
+    lhs_ub = -np.ones_like(b_ub)*np.inf     # LHS of UB constraints is -inf
+    rhs_ub = b_ub                           # RHS of UB constraints is b_ub
+    lhs_eq = b_eq                           # Equality constaint is inequality
+    rhs_eq = b_eq                           # constraint with LHS=RHS
+    lhs = np.concatenate((lhs_ub, lhs_eq))
+    rhs = np.concatenate((rhs_ub, rhs_eq))
+
+    if issparse(A_ub) or issparse(A_eq):
+        A = vstack((A_ub, A_eq))
```

Comment:
I don't see the difference between these branches other than vstack numpyness but didn't understand why.

Reply:
If I remember correctly, we did it this way to avoid some casting issues between scipy.sparse.csc_matrix and coo_matrix.  Doing it this way preserved the csc_matrix.  Maybe @mdhaber can comment on whether this is doing what we think it's doing

Reply:
I'm not sure I understand the question. `vstack` is `scipy.sparse.vstack` and is being used when one of the matrices is sparse. `scipy.sparse.vstack` doesn't work when neither is sparse:
```
import scipy.sparse

import numpy as np

scipy.sparse.vstack(np.eye(2), np.eye(2))
Traceback (most recent call last):

  File "<ipython-input-20-21bbc421af21>", line 1, in <module>
    scipy.sparse.vstack(np.eye(2), np.eye(2))

  File "C:\Users\matth\Desktop\scipy\scipy\sparse\construct.py", line 501, in vstack
    return bmat([[b] for b in blocks], format=format, dtype=dtype)

  File "C:\Users\matth\Desktop\scipy\scipy\sparse\construct.py", line 550, in bmat
    raise ValueError('blocks must be 2-D')

ValueError: blocks must be 2-D
```
Either way, we need to cast to `csc_matrix` afterwards because `np.vstack` returns a numpy array and `scipy.sparse.vstack` returns a COO matrix.

Perhaps we could have converted both `A_ub` and `A_eq` to COO before using `scipy.sparse.vstack`? I haven't tested to see whether that's faster.

Reply:
I missed the sparse import at the top completely :(